### PR TITLE
Implement document load cancellation

### LINKED ISSUE #19309: Implement document load cancellation
This has two parts to it:
* make pressing escape in the default embedding cancel the ongoing load of the foreground, top-most document
* cause all ongoing subresource loads for the document to be cancelled when the document is cancelled

The easiest way to tie the loads together might be to move the cancellation sender into the [LoadType](http://doc.servo.org/script/document_loader/enum.LoadType.html) enum.

Blocked on #19274.

### LINKED ISSUE #21113: Forcibly close EventSources as part of Document unloading
https://html.spec.whatwg.org/multipage/#unloading-documents:concept-eventsource-forcibly-close

Blocked by https://github.com/servo/servo/issues/19309

I might end up doing it as part of 'document aborting'

### LINKED ISSUE #21114: Aborting a document shoud only cancel tasks from the networking task source
Currently, when aborting a document, we cancel all tasks, see https://github.com/servo/servo/blob/c71c55e542c57f6ec4e5e77be750f74f50340e56/components/script/dom/document.rs#L2034

The spec says we shoud be "Cancel any instances of the fetch algorithm in the context of document, discarding any tasks queued for them", which I read as saying 'cancel all networking tasks, not all tasks".  https://html.spec.whatwg.org/multipage/browsing-the-web.html#abort-a-document

This is also necessary to allow for  [failing the connection of an EventSource](https://html.spec.whatwg.org/multipage/server-sent-events.html#fail-the-connection), which happens as part of document aborting. This 'failing the connection' happens via a task on the 'remote event task source', and it should still be processed after the document has been aborted, I think. See https://github.com/servo/servo/issues/21112

I might end up doing this as part of https://github.com/servo/servo/issues/19309

### PR DESCRIPTION
**Pull request by gterzian**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #19309 fix #21114 fix #21113 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21111)

### CODE REVIEW

**Comment by gterzian** [comment]
Not ready for review yet...

**Comment by KiChjang** [comment]
We should probably also use this opportunity to rename cancel to terminate, since that's the exact same term used by the spec.

File: components/script/dom/eventsource.rs [hunk 1]
```diff
@@ -490,13 +494,23 @@ impl EventSource {
...
 }
 
+// https://html.spec.whatwg.org/multipage/#garbage-collection-2
+impl Drop for EventSource {
+    fn drop(&mut self) {
+        // If an EventSource object is garbage collected while its connection is still open,
+        // the user agent must abort any instance of the fetch algorithm opened by this EventSource.
+        self.canceller.borrow_mut().cancel();
+    }
+}
```
Comment (gterzian) [inline components/script/dom/eventsource.rs]: Question on this, is the `Drop` of a Rust `DomRoot` going to be called when the associated `JsObject` is garbage collected in the JS engine?

Comment (jdm) [reply components/script/dom/eventsource.rs]: Yes.

File: components/script/dom/document.rs [hunk 2]
```diff
@@ -2029,14 +2029,18 @@ impl Document {
...
         self.deferred_scripts.clear();
+        let global_scope = self.window.upcast::<GlobalScope>();
+        if self.loader.borrow_mut().cancel_all_loads() || global_scope.close_event_sources() {
+            // If any loads were canceled.
+            self.salvageable.set(false);
+        };
 
-        // TODO: https://github.com/servo/servo/issues/15236
-        self.window.cancel_all_tasks();
+        self.window.cancel_all_tasks_from_source(TaskSourceName::Networking);
```
Comment (KiChjang) [inline components/script/dom/document.rs]: This carries an assumption that all the tasks queued on the networking task source are fetches. I think it is ok for now, but we'd definitely want to leave a comment here clarifying the assumption. Also, step annotations on this algorithm would be nice.

File: components/script/dom/eventsource.rs [hunk 3]
```diff
@@ -390,6 +392,14 @@ impl FetchResponseListener for EventSourceContext {
         }
         self.reestablish_the_connection();
     }
+
+    fn process_response_done(&mut self, aborted:bool) {
+        if aborted {
+            // https://html.spec.whatwg.org/multipage/
+            // server-sent-events.html#sse-processing-model:fail-the-connection
```
Comment (KiChjang) [inline components/script/dom/eventsource.rs]: Wrong link? I got https://html.spec.whatwg.org/multipage/server-sent-events.html#sse-processing-model:fail-the-connection-3 which explains what happens when the fetch is aborted.

Comment (gterzian) [reply components/script/dom/eventsource.rs]: Yep, thanks!

File: components/script/dom/document.rs (hunk 2 above)
Comment (gterzian) [reply components/script/dom/document.rs]: Good catch, I'll add a note.

Comment (gterzian) [reply components/script/dom/document.rs]: I've also filed https://github.com/whatwg/html/issues/3837

**Comment by gterzian** [comment]
```
Error:
/home/servo/buildbot/slave/linux-dev/build/python/_virtualenv/local/lib/python2.7/site-packages/cryptography/hazmat/primitives/constant_time.py:26: CryptographyDeprecationWarning: Support for your Python version is deprecated. The next version of cryptography will remove support. Please upgrade to a 2.7.x release that supports hmac.compare_digest as soon as possible.
  utils.DeprecatedIn23,
awscli 1.15.56 has requirement botocore==1.10.55, but you'll have botocore 1.5.95 which is incompatible.
```

https://github.com/servo/saltfs/issues/862

**Comment by gterzian** [comment]
As well as a whole lot of crashes, I seem to have broken something quite fundamental, I'll switch to local debugging...

**Comment by jdm** [comment]
Is this still WIP?

**Comment by gterzian** [comment]
@jdm I was planning to still implement "make pressing escape in the default embedding cancel the ongoing load", but it might actually be a good idea to review this first, and merge it, and only do the embedder work when the 'internal' part is reviewed and merged. So I've filed https://github.com/servo/servo/issues/21231 and removed WIP from this one.

Ready for review, @mbrubeck did you want to do this one?

File: components/script/dom/globalscope.rs [hunk 4]
```diff
@@ -131,6 +133,9 @@ pub struct GlobalScope {
     /// Vector storing closing references of all workers
     #[ignore_malloc_size_of = "Arc"]
     list_auto_close_worker: DomRefCell<Vec<AutoCloseWorker>>,
+
+    /// Vector storing references of all eventsources.
+    event_sources: DomRefCell<Vec<DomRoot<EventSource>>>,
```
Comment (jdm) [inline components/script/dom/globalscope.rs]: The event sources will now never be GCed while they're in this vector. Instead, if we store a WeakRefVec and add the weakReferenceable attribute to the EventSource interface in Bindings.conf, we can ensure that these references don't keep the underlying objects alive.

**Review by jdm — CHANGES_REQUESTED** [review]
These changes look like a good improvement overall!

File: components/script/dom/globalscope.rs (hunk 4 above)
Comment (gterzian) [reply components/script/dom/globalscope.rs]: Thanks! 

@KiChjang that seems that the answer we were looking for in https://github.com/servo/servo/pull/21218 as well...

Comment (KiChjang) [reply components/script/dom/globalscope.rs]: @gterzian Yes, I had a suspicion that DOM struct field members will not be GC'd, so we'd definitely need to pop the elements off of the Vec somehow...

**Comment by gterzian** [comment]
@jdm please take another look at https://github.com/servo/servo/pull/21111/commits/36247aa81bb12988a212d681543f28f23e1d760c
I've largely been "inspired" by [WeakMediaQueryListVec](https://github.com/servo/servo/blob/4c5a7eaa14c48daa3bf1f40a65e55b9a97a01b5f/components/script/dom/mediaquerylist.rs#L120)

**Review by jdm — APPROVED** [review]
Sensible choice! We'll merge this as soon as our CI is working again.

**Comment by jdm** [comment]
@bors-servo retry
* builder was full

**Comment by gterzian** [comment]
Looks like websockets timeout, and a possible intermittent PASS that I will investigate:

```

{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Verifies the resolution of performance.now() is at least 20 microseconds.", "test": "/performance-timeline/webtiming-resolution.any.worker.html", "line": 41625, "action": "test_result", "expected": "FAIL"}
```

**Comment by jdm** [comment]
@bors-servo retry
* #21276

**Comment by gterzian** [comment]
@jdm The `PASS` for `/performance-timeline/webtiming-resolution.any.worker.html` proved consistent on my machine, so I've update the expectation. I added to the first commit, because while I suspected it might be related to my changes to the parser code, this actually wasn't the case. So I can only say "something as part of, I think, the first commit, made this test pass".

**Comment by jdm** [comment]
Interesting. It seems implausible that it would be related to these changes, but clearly I'm incorrect!
@bors-servo r+

**Comment by gterzian** [comment]
Oh well....

```
{"status": "FAIL", "group": "default", "message": "assert_greater_than_equal: The second performance.now() should be much greater than the first expected a number greater than or equal to 0.02 but got 0.012958000000001135", "stack": "testTimeResolution/<@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:8:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20\ntest@http://web-platform.test:8000/resources/testharness.js:545:9\ntestTimeResolution@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:2:5\n@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:23:1\n@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.worker.js:8:1\n", "subtest": "Verifies the resolution of performance.now() is at least 20 microseconds.", "test": "/performance-timeline/webtiming-resolution.any.worker.html", "line": 104072, "action": "test_result", "expected": "PASS"}
```

**Comment by gterzian** [comment]
Previous change reverted... @jdm

**Comment by jdm** [comment]
@bors-servo retry
* #21276

**Comment by gterzian** [comment]
@jdm thanks, I think it needs an r+ because I rebased in the meantime...

**Comment by jdm** [comment]
More build machine disk space woes.

**Comment by gterzian** [comment]
This one has come back, and this time it's OK...
```
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/scripting-1/the-script-element/script-onerror-insertion-point-2.html", "line": 94095, "action": "test_result", "expected": "TIMEOUT"}
```

**Comment by gterzian** [comment]
Ok, removed the `TIMEOUT` for `{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/scripting-1/the-script-element/script-onerror-insertion-point-2.html", "line": 94095, "action": "test_result", "expected": "TIMEOUT"`

**Pull request merged**